File: python-stt/diarization.py
"""
Speech-to-Text with Speaker Diarization Example

Demonstrates STT with speaker identification via WebSocket.
Diarization identifies different speakers in an audio stream.

Usage:
    python diarization.py
"""

import asyncio
import json
import os

import socketio
from dotenv import load_dotenv

load_dotenv()

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

CONFIG = {
    "api_host": os.getenv("API_HOST", "https://sdk.verbum.ai"),
    "api_key": os.getenv("API_KEY", ""),
    "language": os.getenv("LANGUAGE", "en-US"),
    "audio_file": os.getenv("AUDIO_FILE", "./sample.wav"),
    "output_file": "output-diarization.json",
}

if not CONFIG["api_key"] or CONFIG["api_key"] == "your_api_key_here":
    print("ERROR: Set API_KEY in your .env file")
    exit(1)


# ---------------------------------------------------------------------------
# Diarization STT Client
# ---------------------------------------------------------------------------


class DiarizationSTTClient:
    def __init__(self, config):
        self.config = config
        self.sio = None
        self.segments = []
        self.connected = False
        self.done = asyncio.Event()

    async def connect(self):
        """Connect to the WebSocket server"""
        print(f"\nConnecting to {self.config['api_host']}/speech...")

        self.sio = socketio.AsyncClient()

        @self.sio.event
        async def connect():
            self.connected = True
            print("✓ Connected")

        @self.sio.on("transcriptionUpdate")
        async def on_update(data):
            if "diarization" in data and "speaker" in data["diarization"]:
                speaker = f"Speaker-{data['diarization']['speaker']}"
                print(f"  [{speaker}] {data.get('transcript', '')}")

        @self.sio.on("transcriptionCompleted")
        async def on_completed(data):
            # Store segments with speaker information
            if "segments" in data:
                for segment in data["segments"]:
                    self.segments.append(
                        {
                            "speaker": f"Speaker-{segment.get('speakerId', '0')}",
                            "transcript": segment.get("transcript"),
                            "startTime": segment.get("startTime"),
                            "endTime": segment.get("endTime"),
                            "confidence": segment.get("confidence"),
                            "speakerId": segment.get("speakerId"),
                        }
                    )
                    print(
                        f"  [Speaker-{segment.get('speakerId', '0')}] "
                        f"{segment.get('transcript')} "
                        f"({segment.get('startTime', 0)}s - {segment.get('endTime', 0)}s)"
                    )

            # Store full transcript
            self.segments.append(
                {
                    "type": "full",
                    "transcript": data.get("transcript"),
                    "timestamp": str(__import__("datetime").datetime.now().isoformat()),
                }
            )

            self.done.set()

        @self.sio.on("transcriptionError")
        async def on_error(data):
            error_msg = data.get("message", "Unknown error") if isinstance(data, dict) else str(data)
            print(f"Transcription error: {error_msg}")
            self.done.set()

        @self.sio.event
        async def disconnect():
            self.connected = False

        try:
            await self.sio.connect(
                f"{self.config['api_host']}/speech",
                auth={"token": self.config["api_key"]},
                transports=["websocket"],
                headers={"language": self.config["language"], "usage": "browser"},
            )
        except Exception as e:
            print(f"Connection failed: {e}")
            raise

    async def start_transcription(self):
        """Start diarization transcription"""
        print(f"\nStarting diarization from: {self.config['audio_file']}")

        if not os.path.exists(self.config["audio_file"]):
            raise FileNotFoundError(f"Audio file not found: {self.config['audio_file']}")

        with open(self.config["audio_file"], "rb") as f:
            audio_data = f.read()

        # Start transcription with diarization enabled
        self.sio.emit(
            "startTranscription",
            {
                "language": [self.config["language"]],
                "encoding": "PCM",
                "sampleRate": 16000,
                "diarization": True,
                "diarizationSpeakerCount": None,  # Auto-detect speaker count
            },
        )

        chunk_size = 1024
        offset = 0

        try:
            while offset < len(audio_data):
                chunk = audio_data[offset : offset + chunk_size]
                offset += chunk_size
                self.sio.emit("audioStream", chunk)
                await asyncio.sleep(0.02)

            self.sio.emit("endStream")
            await asyncio.wait_for(self.done.wait(), timeout=300)
        except asyncio.TimeoutError:
            print("Transcription timeout")
            raise

    async def disconnect(self):
        """Disconnect from the server"""
        if self.sio:
            await self.sio.disconnect()
            print("Disconnected")

    def save_results(self):
        """Save results to JSON file"""
        # Group segments by speaker
        speaker_map = {}
        segment_list = []

        for segment in self.segments:
            if segment.get("type") != "full":
                speaker = segment["speaker"]
                if speaker not in speaker_map:
                    speaker_map[speaker] = []
                speaker_map[speaker].append(segment)
            segment_list.append(segment)

        results = {
            "config": {
                "language": self.config["language"],
                "audioFile": os.path.basename(self.config["audio_file"]),
                "diarization": True,
            },
            "timestamp": str(__import__("datetime").datetime.now().isoformat()),
            "summary": {
                "totalSpeakers": len(speaker_map),
                "speakers": list(speaker_map.keys()),
            },
            "speakerBreakdown": speaker_map,
            "allSegments": segment_list,
        }

        with open(self.config["output_file"], "w") as f:
            json.dump(results, f, indent=2)

        print(f"\nResults saved to: {self.config['output_file']}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


async def main():
    client = DiarizationSTTClient(CONFIG)

    try:
        await client.connect()
        await client.start_transcription()
        client.save_results()
    except Exception as e:
        print(f"Fatal error: {e}")
        exit(1)
    finally:
        await client.disconnect()


if __name__ == "__main__":
    asyncio.run(main())

File: nodejs-stt/diarization.js
/**
 * Speech-to-Text with Speaker Diarization Example
 *
 * Demonstrates STT with speaker identification via WebSocket.
 * Diarization identifies different speakers in an audio stream.
 *
 * Usage:
 *   npm run diarization
 */

'use strict';

require('dotenv').config();
const fs = require('fs');
const path = require('path');
const { io } = require('socket.io-client');

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const CONFIG = {
  apiHost: process.env.API_HOST || 'https://sdk.verbum.ai',
  apiKey: process.env.API_KEY || '',
  language: process.env.LANGUAGE || 'en-US',
  audioFile: process.env.AUDIO_FILE || path.join(__dirname, './sample.wav'),
  outputFile: path.join(__dirname, 'output-diarization.json'),
};

if (!CONFIG.apiKey || CONFIG.apiKey === 'your_api_key_here') {
  console.error('ERROR: Set API_KEY in your .env file');
  process.exit(1);
}

// ---------------------------------------------------------------------------
// Diarization STT Client
// ---------------------------------------------------------------------------

class DiarizationSTTClient {
  constructor(config) {
    this.config = config;
    this.socket = null;
    this.segments = [];
    this.currentSegment = null;
  }

  /**
   * Connect to the WebSocket server
   */
  async connect() {
    return new Promise((resolve, reject) => {
      console.log(`\nConnecting to ${this.config.apiHost}/listen...`);

      this.socket = io(`${this.config.apiHost}/listen`, {
        query: {
          token: this.config.apiKey,
          language: this.config.language,
          usage: 'browser',
        },
        transports: ['websocket'],
      });

      this.socket.on('connect', () => {
        console.log('✓ Connected');
        resolve();
      });

      this.socket.on('connect_error', (error) => {
        console.error('Connection error:', error.message);
        reject(error);
      });

      // Listen for transcription results with diarization info
      this.socket.on('speechRecognized', (data) => {
        const status = data.status || '';
        if (status === 'recognizing') {
          if (data.speakerId) {
            const speaker = `Speaker-${data.speakerId}`;
            console.log(`  [${speaker}] ${data.text}`);
          }
        } else if (status === 'recognized') {
          if (data.segments) {
            // Store segments with speaker information
            data.segments.forEach((segment) => {
              this.segments.push({
                speaker: `Speaker-${segment.speakerId || '0'}`,
                transcript: segment.transcript,
                startTime: segment.startTime,
                endTime: segment.endTime,
                confidence: segment.confidence,
                speakerId: segment.speakerId,
              });
              console.log(
                `  [Speaker-${segment.speakerId || '0'}] ${segment.transcript} (${segment.startTime}s - ${
                  segment.endTime
                }s)`
              );
            });
          }

          // Also store the full transcript
          this.segments.push({
            type: 'full',
            transcript: data.text,
            timestamp: new Date().toISOString(),
          });
        }
      });

      this.socket.on('transcriptionError', (error) => {
        console.error('Transcription error:', error);
        reject(new Error(error.message || 'Transcription failed'));
      });
    });
  }

  /**
   * Start diarization transcription
   */
  async startTranscription() {
    console.log(`\nStarting diarization from: ${this.config.audioFile}`);

    if (!fs.existsSync(this.config.audioFile)) {
      throw new Error(`Audio file not found: ${this.config.audioFile}`);
    }

    const audioData = fs.readFileSync(this.config.audioFile);
    const chunkSize = 1024;

    return new Promise((resolve, reject) => {
      let offset = 0;

      // Start transcription with diarization enabled
      this.socket.emit('startTranscription', {
        language: [this.config.language],
        encoding: 'PCM',
        sampleRate: 16000,
        diarization: true, // Enable speaker diarization
        diarizationSpeakerCount: null, // Auto-detect speaker count, or set to specific number (2, 3, etc.)
      });

      const streamInterval = setInterval(() => {
        if (offset >= audioData.length) {
          clearInterval(streamInterval);
          this.socket.emit('streamEnd');
          resolve();
          return;
        }

        const chunk = audioData.slice(offset, offset + chunkSize);
        offset += chunkSize;

        this.socket.emit('audioStream', chunk);
      }, 20);

      const timeout = setTimeout(() => {
        clearInterval(streamInterval);
        reject(new Error('Transcription timeout'));
      }, 300000);

      this.socket.once('speechRecognized', (data) => {
        if (data.status === 'recognized') {
          clearTimeout(timeout);
        }
      });
    });
  }

  /**
   * Disconnect from the server
   */
  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      console.log('Disconnected');
    }
  }

  /**
   * Save results to JSON file
   */
  saveResults() {
    // Group segments by speaker for easier reading
    const speakerMap = {};
    const segmentList = [];

    this.segments.forEach((segment) => {
      if (segment.type !== 'full') {
        if (!speakerMap[segment.speaker]) {
          speakerMap[segment.speaker] = [];
        }
        speakerMap[segment.speaker].push(segment);
      }
      segmentList.push(segment);
    });

    const results = {
      config: {
        language: this.config.language,
        audioFile: path.basename(this.config.audioFile),
        diarization: true,
      },
      timestamp: new Date().toISOString(),
      summary: {
        totalSpeakers: Object.keys(speakerMap).length,
        speakers: Object.keys(speakerMap),
      },
      speakerBreakdown: speakerMap,
      allSegments: segmentList,
    };

    fs.writeFileSync(this.config.outputFile, JSON.stringify(results, null, 2));
    console.log(`\nResults saved to: ${this.config.outputFile}`);
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const client = new DiarizationSTTClient(CONFIG);

  try {
    await client.connect();
    await client.startTranscription();
    client.saveResults();
  } catch (error) {
    console.error('Fatal error:', error.message);
    process.exit(1);
  } finally {
    client.disconnect();
  }
}

main();
